projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-tool/tool.cpp

```diff
@@ -528,19 +533,139 @@ cntrl_tracing_callback(rocprofiler_callback_tracing_record_t record,
+    for(auto ctx : *ctxs)
     {
-        if(record.phase == ROCPROFILER_CALLBACK_PHASE_ENTER &&
-           record.operation == ROCPROFILER_MARKER_CONTROL_API_ID_roctxProfilerPause)
+        if(int active = 0;
+           ctx != null_context_id &&
+           rocprofiler_context_is_active(ctx, &active) == ROCPROFILER_STATUS_SUCCESS && active != 0)
+            _active_contexts++;
+    }
+
+    auto _first_pause_resume = false;
+
+    // setup once flag to track first pause/resume call
+    static auto _once_flag = std::once_flag{};
+    std::call_once(_once_flag, [&]() { _first_pause_resume = true; });
+
+    if(ctxs && record.kind == ROCPROFILER_CALLBACK_TRACING_MARKER_CONTROL_API)
+    {
+        if(!tool::get_config().collection_periods.empty())
         {
-            ROCPROFILER_CALL(rocprofiler_stop_context(*ctx), "pausing context");
+            ROCP_WARNING_IF(record.phase == ROCPROFILER_CALLBACK_PHASE_ENTER)
+                << "rocprofv3 collection period(s) enabled, ignoring roctxProfilerPause/Resume";
+            return;
+        }
+        else if(record.phase == ROCPROFILER_CALLBACK_PHASE_ENTER &&
+                record.operation == ROCPROFILER_MARKER_CONTROL_API_ID_roctxProfilerPause)
+        {
+            // provide warning if thread id is used since rocprofv3 does not support thread-local
+            // pause/resume
+            if(auto* _data =
+                   static_cast<rocprofiler_callback_tracing_marker_api_data_t*>(record.payload);
+               _data && _data->args.roctxProfilerPause.tid != 0)
+            {
+                ROCP_INFO_IF(_data->args.roctxProfilerPause.tid !=
+                             static_cast<rocprofiler_thread_id_t>(getpid()))
+                    << fmt::format("roctxProfilerResume(tid={}) invoked on thread {}. rocprofv3 "
+                                   "does not support thread-local pause/resume (only global "
+                                   "pause/resume). Use tid=0 or only call from main thread ({}).",
+                                   _data->args.roctxProfilerPause.tid,
+                                   common::get_tid(),
+                                   getpid());
+            }
+
+            // if using the selected regions reference counting, only pause when the count goes to
+            // zero
+            auto _ref_count =
+                (tool::get_config().selected_regions_ref_count) ? pause_resume_count-- : int64_t{0};
+
+            if(tool::get_config().selected_regions && _first_pause_resume && _ref_count < 0 &&
+               _active_contexts == 0)
+            {
+                ROCP_CI_LOG(INFO)
+                    << "first call to roctxProfilerPause ignored for selected regions profiling "
+                       "with the selected regions reference counting mode enabled";
+                // increment so that it doesn't trigger else if(_ref_count != 0) block below
+                _ref_count = ++pause_resume_count;
+            }
```

Comment:
```suggestion
            int64_t _ref_count = 0;
            
            if (!_first_pause_resume)
            {
                _ref_count = (tool::get_config().selected_regions_ref_count) ? --pause_resume_count : int64_t{0};
            }
            else if (_first_pause_resume && tool::get_config().selected_regions && tool::get_config().selected_regions_ref_count)
            {
                ROCP_CI_LOG(INFO)
                    << "first call to roctxProfilerPause ignored for selected regions profiling "
                       "with the selected regions reference counting mode enabled";
            }
```

I took an attempt at simplifying the logic here. I wanted to avoid updating the atomic unnecessarily.
I also changed the post-decrement to a pre-decrement, which makes the logic match the comment on line 592, i.e. when the ref_count becomes 0, a pause occurs. 

This suggestion pairs with a suggestion for the resume logic, so please pay attention to that comment as well.